### PR TITLE
Development

### PR DESCRIPTION
**Pull request by garciajordy**
**Enumerable Methods List:**

- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject ( we create a my_inject that takes + and * blocks so we didn't create multiply_els as it is a part of the my_inject)

### CODE REVIEW

**Comment by mimipeshy** [comment]
#### STATUS: CHANGES REQUIRED :red_circle: :partying_face:
Hello team @garciajordy  and @JulianaOsemeke    👋🏾 👋🏾 great job on the first submission of this project

#### Highlights :white_check_mark: :white_check_mark:

- [x]  Well written README file, simple to follow and understand
- [x]  Clean well written code

However, before marking the project as complete, a few changes need to be made. Check them out in the comments ⬇️ 

- [x] You are missing the `multiply_els` function which accepts an array as an argument and multiplies all the elements of the array together using #my_inject
     

That's all :smile: kindly submit another review when done.

_Feel free to reach out to me on slack if you have any questions or something is unclear, or just comment on this PR and tag my name  (@mimipeshy) and I will be glad to help_
Regards
Peris Ndanu

Regards
Peris Ndanu

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# enumerable.rb
+module Enumerable
+  def my_each
+    i = 0
+    while to_a.length > i
+      yield(to_a[i])
+      i += 1
+    end
+  end
+
```
Comment (mimipeshy) [inline enumerables.rb:10]: 1. enumerables #my_each should return an Enumerator if no block is given
2.  enumerables #my_each when a block is given and self is an array it should return the array itself after calling the given block once for each element in self
3.  enumerables #my_each when a block is given when self is a range it should return the range itself after calling the given block once for each element in self
4.

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  def my_each_with_index
+    i = 0
+    while to_a.length > i
+      yield(to_a[i], i)
+      i += 1
+    end
+  end
+
```
Comment (mimipeshy) [inline enumerables.rb:24]: 1. enumerables #my_each_with_index returns an enumerator if no block is given
     Failure/Error: expect(array.my_each_with_index).to be_an(Enumerator)
2. enumerables #my_each_with_index when a block is given and self is an array it should return the array itself after calling the given block once for each element in self
   Failure/Error: expect(array.my_each_with_index(&block)).to eq(array.each_with_index(&block))
3. enumerables #my_each_with_index when a block is given when self is a range, it should return the range itself after calling the given block once for each element in self
Failure/Error: expect(range.my_each_with_index(&block)).to eq(range.each_with_index(&block))

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  def my_select
+    new_array = []
+    to_a.my_each { |item| new_array.push(item) if yield(item) }
+    new_array
+  end
+
```
Comment (mimipeshy) [inline enumerables.rb:32]: 1. enumerables #my_select returns an enumerator if no block is given

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  def my_all?
+    to_a.my_each { |item| return false if yield(item) == false }
+    true
+  end
+
```
Comment (mimipeshy) [inline enumerables.rb:47]: 1. enumerables #my_all? when no block or argument is given returns true when none of the collection members are false or nil
Failure/Error: expect(true_array.my_all?).to be true_array.all?

2. enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil
Failure/Error: expect(false_array.my_all?).to be false_array.all?

3. enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
Failure/Error: expect(array.my_all?(Integer)).to be array.all?(Integer)

4. enumerables #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
Failure/Error: expect(numbers.my_all?(Numeric)).to be numbers.all?(Numeric)

5.  enumerables #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
Failure/Error: expect(array.my_all?(Integer)).to be array.all?(Integer)

6.  enumerables #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
Failure/Error: expect(words.my_all?(/d/)).to be words.all?(/d/)

7. enumerables #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
Failure/Error: expect(words.my_all?(/o/)).to be words.all?(/o/)

8. enumerables #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
Failure/Error: expect(array.my_all?(3)).to be array.all?(3)

9. enumerables #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
Failure/Error: expect(array.my_all?(3)).to be array.all?(3)

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  def my_any?
+    to_a.my_each { |item| return true if yield(item) }
+    false
+  end
```
Comment (mimipeshy) [inline enumerables.rb:62]: 1. enumerables #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
Failure/Error: expect(true_array.my_any?).to be true_array.any?
2. enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true
 Failure/Error: expect(false_array.my_any?).to be false_array.any?
3. enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
Failure/Error: expect(array.my_any?(Numeric)).to be array.any?(Numeric)
4. enumerables #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
Failure/Error: expect(words.my_any?(Integer)).to be words.any?(Integer)
5.  enumerables #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
 Failure/Error: expect(words.my_any?(/d/)).to be words.any?(/d/)
6. enumerables #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
Failure/Error: expect(words.my_any?(/z/)).to be words.any?(/z/)
7. enumerables #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
Failure/Error: expect(words.my_any?('cat')).to be words.any?('cat')
8.  enumerables #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
Failure/Error: expect(words.my_any?('cat')).to be words.any?('cat')

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_none?
+    to_a.my_each { |item| return false if yield(item) }
+    true
+  end
```
Comment (mimipeshy) [inline enumerables.rb:70]: - [ ]  enumerables #my_none? when no block or argument is given returns true only if none of the collection members is true
      Failure/Error: expect(false_array.my_none?).to be true

- [ ] enumerables #my_none? when no block or argument is given returns false only if one of the collection members is true
      Failure/Error: expect(true_array.my_none?).to be false
- [ ] enumerables #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
      Failure/Error: expect(array.my_none?(String)).to be array.my_none?(String)

- [ ]  enumerables #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
      Failure/Error: expect(array.my_none?(Numeric)).to be array.my_none?(Numeric)

- [ ] enumerables #my_none? when a class is passed as an argument returns false if any of the collection is a member of such class
      Failure/Error: expect(array.my_none?(String)).to be array.my_none?(String)

- [x] enumerables #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
      Failure/Error: expect(words.my_none?(/z/)).to be words.none?(/z/)

- [x]  enumerables #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
      Failure/Error: expect(words.my_none?(/d/)).to be words.none?(/d/)

- [x] enumerables #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
      Failure/Error: expect(words.my_none?(5)).to be words.none?(5)

- [x]  enumerables #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
      Failure/Error: expect(words.my_none?(5)).to be words.none?(5)

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_count?
+    i = 0
+    to_a.my_each { |_item| i += 1 }
+    i
+  end
```
Comment (mimipeshy) [inline enumerables.rb:82]: - [x] enumerables #my_count returns the number of items in enum through enumeration
      Failure/Error: expect(array.my_count).to eq array.count

- [x] enumerables #my_count returns the number of items in enum through enumeration::range
      Failure/Error: expect(range.my_count).to eq range.count

- [x] enumerables #my_count counts the number of items in enum that are equal to item if an argument is given
      Failure/Error: expect(array.my_count(LOWEST_VALUE)).to eq array.count(LOWEST_VALUE)

- [x] enumerables #my_count counts the number of elements yielding a true value if a block is given
      Failure/Error: expect(array.my_count(&block)).to eq array.count(&block)

- [x] enumerables #my_count does not mutate the original array
      Failure/Error: array.my_count { |num| num + 1 }

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  def my_map(proc = nil)
+    new_array = []
+    if proc.nil?
+      to_a.my_each { |item| new_array << yield(item) }
+
+    else
+      to_a.my_each { |item| new_array << proc.call(item) }
+    end
+    new_array
+  end
```
Comment (mimipeshy) [inline enumerables.rb:95]: - [x] enumerables #my_map returns an Enumerator if no block is given
      Failure/Error: expect(array.my_map).to be_an(Enumerator)

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+  def my_inject
+    sum = 0
+    to_a.my_each { |item| sum = yield(sum, item) }
+    return sum if sum != 0
+
+    sum = 1
+    to_a.my_each { |item| sum = yield(sum, item) }
+    sum
+  end
```
Comment (mimipeshy) [inline enumerables.rb:65]: - [ ] enumerables #my_inject searches for the longest word in an array of strings
      Failure/Error: let(:search) { proc { |memo, word| memo.length > word.length ? memo : word } }

- [ ] enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
      Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }

- [ ] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
      Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)

- [ ]  enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
      Failure/Error: expect(range.my_inject(:*)).to eq range.inject(:*)

- [ ] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
      Failure/Error: expect(array.my_inject(20, :*)).to eq array.inject(20, :*)

- [ ] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
      Failure/Error: expect(range.my_inject(2, :*)).to eq range.inject(2, :*)

Comment (mimipeshy) [inline enumerables.rb:108]: - [x] enumerables #my_inject searches for the longest word in an array of strings
      Failure/Error: let(:search) { proc { |memo, word| memo.length > word.length ? memo : word } }

- [x] enumerables #my_inject when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
      Failure/Error: actual = range.my_inject(4) { |prod, n| prod * n }

- [x] enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
      Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)

- [x]  enumerables #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
      Failure/Error: expect(range.my_inject(:*)).to eq range.inject(:*)

- [x] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
      Failure/Error: expect(array.my_inject(20, :*)).to eq array.inject(20, :*)

- [x] enumerables #my_inject when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
      Failure/Error: expect(range.my_inject(2, :*)).to eq range.inject(2, :*)

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,66 @@
+# enumerable.rb
+module Enumerable
+  def my_each
+    i = 0
+    while to_a.length > i
+      yield(to_a[i])
+      i += 1
+    end
+  end
```
Comment (mimipeshy) [inline enumerables.rb:12]: - [x] enumerables #my_each returns an Enumerator if no block is given
     Failure/Error: expect(array.my_each).to be_an(Enumerator)

- [x] enumerables #my_each when a block is given when self is an array returns the array itself after calling the given block once for each element in self
     Failure/Error: expect(array.my_each(&block)).to eq(array.each(&block))

- [x] enumerables #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self
     Failure/Error: expect(range.my_each(&block)).to eq(range.each(&block))

- [x] enumerables #my_each_with_index returns an enumerator if no block is given
     Failure/Error: expect(array.my_each_with_index).to be_an(Enumerator)

File: enumerables.rb (hunk 2 above)
Comment (garciajordy) [reply enumerables.rb:24]: Solved all the requested changes

File: enumerables.rb (hunk 3 above)
Comment (garciajordy) [reply enumerables.rb:32]: Solved the Requested change

File: enumerables.rb (hunk 4 above)
Comment (garciajordy) [reply enumerables.rb:47]: Solved all the requested changes and also tested it

File: enumerables.rb (hunk 5 above)
Comment (garciajordy) [reply enumerables.rb:62]: Solved all the requested changes and also tested it

File: enumerables.rb (hunk 6 above)
Comment (garciajordy) [reply enumerables.rb:70]: Everything is done!!! Sorry for not checking all the boxes but I am unable to check all of them on GitHub, don't know why though

File: enumerables.rb (hunk 7 above)
Comment (garciajordy) [reply enumerables.rb:82]: Solved all the requested changes and also tested it

File: enumerables.rb (hunk 8 above)
Comment (garciajordy) [reply enumerables.rb:95]: Solved the requested change

File: enumerables.rb (hunk 10 above)
Comment (garciajordy) [reply enumerables.rb:12]: Solved all the requested changes

**Comment by garciajordy** [comment]
We worked together on resolving all the Linters issues as there were only minor problems we decided to have 1 navigator and 1 driver

**Review by rukundoeric — CHANGES_REQUESTED** [review]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve 
  - [x] [MANDATORY] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
    - when array is self and no block given `[1, 2i, 3.14].my_all(Numeric)`
      - Expected: `true` 
      - Got: `false`
  - [x] [MANDATORY] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
    - when array is self and no block given `[1, 2, 3].my_any(Numeric)`
      - Expected: `true` 
      - Got: `false`
  - [x] [MANDATORY] #my_none? returns false if the block ever returns true for all elements
    - when array is self and no block given `[1, 2, 3, 10].my_none?(&proc { |num| num > 9 })`
      - Expected: `false` 
      - Got: `true`
  - [x] [MANDATORY] #my_none? returns false if the block ever returns true for all elements::range
    - when array is self and no block given `Range.new(5, 50).my_none?(&proc { |num| num > 9 })`
      - Expected: `false` 
      - Got: `true`
  - [x] [MANDATORY] #my_none? when no block or argument is given returns false only if one of the collection members is true
    - Example: `[1, true, 'hi', []].my_none?`
      - Expected: `true` 
      - Got: `false`

_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Juliana Osemeke & Jordy Garcia** 
Good job with your first submission for this project :muscle:  :muscle:
Your Enumerable methods passes all the tests :clap: :clap: :clap:  
There are just a couple of details to improve before you can go ahead:

- [x] You are not following the correct git-flow for this project, development should be your main branch and you should have a feature (or some other descriptive name) branch to make the pull request from feature to development.
- [x] Don't overuse rubocop disable
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Pull request merged**